### PR TITLE
resholve: add usable (overridden) python27

### PR DESCRIPTION
**Pull request by abathur**
###### Description of changes

Minimum-viable-change to keep resholve (and all packages using resholve's Nix API) building after #201859.

I've also added an ~unrelated commit to clean up some old resholve aliases that just happen to be ripe for removal today.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by cole-h — DISMISSED** [review]
Should this be a draft / otherwise noted as "not suitable for merging" (yet) because it includes https://github.com/NixOS/nixpkgs/pull/201859? (It would be a shame if this merged and bypassed whatever conversation / review may be happening in that PR.)

EDIT: "requested changes" just so that it's not assumed to be good-for-merging until I get some clarification on my suggestion :P

**Review by thiagokokada — COMMENTED** [review]
Ugly, but this works to avoid adding the custom `python27` in `all-packages.nix`:

```patch
diff --git a/pkgs/development/misc/resholve/default.nix b/pkgs/development/misc/resholve/default.nix
index 771b85e177a..bc6c7270089 100644
--- a/pkgs/development/misc/resholve/default.nix
+++ b/pkgs/development/misc/resholve/default.nix
@@ -1,16 +1,28 @@
-{ callPackage
-, deNeuteredPy27OnlyForResholve
+{ lib
+, pkgs
+, pkgsBuildHost
 , ...
 }:
 
 let
+  python27' = (pkgs.python27.overrideAttrs (old:
+    {
+      # Overriding `meta.knownVulnerabilities` here, see #201859 for why it exists
+      # In resholve case this will not be a security issue,
+      # since the code should always run inside nix sandbox (when enabled)
+      meta = old.meta // { knownVulnerabilities = [ ]; };
+    }
+  )).override {
+    self = python27';
+    pkgsBuildHost = pkgsBuildHost // { python27 = python27'; };
+  };
+  callPackage = lib.callPackageWith (pkgs // { python27 = python27'; });
   source = callPackage ./source.nix { };
-  deps = callPackage ./deps.nix { python27 = deNeuteredPy27OnlyForResholve; };
+  deps = callPackage ./deps.nix { };
 in
 rec {
   # resholve itself
   resholve = callPackage ./resholve.nix {
-    python27 = deNeuteredPy27OnlyForResholve;
     inherit (source) rSrc version;
     inherit (deps.oil) oildev;
     inherit resholve-utils;
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 86cde56da9c..e24d1e9130d 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -15868,10 +15868,6 @@ with pkgs;
     pythonAttr = "python2Full";
     x11Support = true;
   };
-  deNeuteredPy27OnlyForResholve = (python27.overrideAttrs(old: { meta = old.meta // { knownVulnerabilities = []; };})).override {
-    self = deNeuteredPy27OnlyForResholve;
-    pythonAttr = "deNeuteredPy27OnlyForResholve";
-  };
   python27Full = python27.override {
     self = python27Full;
     pythonAttr = "python27Full";
```

**Comment by thiagokokada** [comment]
BTW, cross-compilation is not working:

```
$ nix build .#pkgsCross.aarch64-multiplatform.mons
error: Package ‘python-2.7.18’ in /nix/store/4ns3zd77y2wpmsrraakp6237388pxpvf-source/pkgs/development/interpreters/python/cpython/2.7/default.nix:329 is marked as insecure, refusing to evaluate.


       Known issues:
        - Python 2.7 has reached its end of life after 2020-01-01. See https://www.python.org/doc/sunset-python-2/.

       You can install it anyway by allowing this package, using the
       following methods:

       a) To temporarily allow all insecure packages, you can use an environment
          variable for a single invocation of the nix tools:

            $ export NIXPKGS_ALLOW_INSECURE=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) for `nixos-rebuild` you can add ‘python-2.7.18’ to
          `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
          like so:

            {
              nixpkgs.config.permittedInsecurePackages = [
                "python-2.7.18"
              ];
            }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
          ‘python-2.7.18’ to `permittedInsecurePackages` in
          ~/.config/nixpkgs/config.nix, like so:

            {
              permittedInsecurePackages = [
                "python-2.7.18"
              ];
            }
(use '--show-trace' to show detailed location information)
```

This is independent if my patch is applied or not. I am assuming there is something in the cross-compilation bootstrap depending on `python27`.

**Comment by thiagokokada** [comment]
A more complete patch:

```patch
diff --git a/pkgs/development/misc/resholve/default.nix b/pkgs/development/misc/resholve/default.nix
index 771b85e177a..855e57c8592 100644
--- a/pkgs/development/misc/resholve/default.nix
+++ b/pkgs/development/misc/resholve/default.nix
@@ -1,16 +1,37 @@
-{ callPackage
-, deNeuteredPy27OnlyForResholve
+{ lib
+, pkgs
+, pkgsBuildBuild
+, pkgsBuildHost
+, pkgsBuildTarget
+, pkgsHostHost
+, pkgsTargetTarget
 , ...
 }:
 
 let
+  python27' = (pkgs.python27.overrideAttrs (old:
+    {
+      # Overriding `meta.knownVulnerabilities` here, see #201859 for why it exists
+      # In resholve case this will not be a security issue,
+      # since the code should always run inside nix sandbox (when enabled)
+      meta = old.meta // { knownVulnerabilities = [ ]; };
+    }
+  )).override {
+    self = python27';
+    # Need to override all targets, otherwise we break cross-compilation
+    pkgsBuildBuild = pkgsBuildBuild // { python27 = python27'; };
+    pkgsBuildHost = pkgsBuildHost // { python27 = python27'; };
+    pkgsBuildTarget = pkgsBuildTarget // { python27 = python27'; };
+    pkgsHostHost = pkgsHostHost // { python27 = python27'; };
+    pkgsTargetTarget = pkgsTargetTarget // { python27 = python27'; };
+  };
+  callPackage = lib.callPackageWith (pkgs // { python27 = python27'; });
   source = callPackage ./source.nix { };
-  deps = callPackage ./deps.nix { python27 = deNeuteredPy27OnlyForResholve; };
+  deps = callPackage ./deps.nix { };
 in
 rec {
   # resholve itself
   resholve = callPackage ./resholve.nix {
-    python27 = deNeuteredPy27OnlyForResholve;
     inherit (source) rSrc version;
     inherit (deps.oil) oildev;
     inherit resholve-utils;
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 86cde56da9c..e24d1e9130d 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -15868,10 +15868,6 @@ with pkgs;
     pythonAttr = "python2Full";
     x11Support = true;
   };
-  deNeuteredPy27OnlyForResholve = (python27.overrideAttrs(old: { meta = old.meta // { knownVulnerabilities = []; };})).override {
-    self = deNeuteredPy27OnlyForResholve;
-    pythonAttr = "deNeuteredPy27OnlyForResholve";
-  };
   python27Full = python27.override {
     self = python27Full;
     pythonAttr = "python27Full";
```

**Comment by thiagokokada** [comment]
Working patch with cross-compilation:

```patch
diff --git a/pkgs/development/misc/resholve/default.nix b/pkgs/development/misc/resholve/default.nix
index 771b85e177a..76edf927769 100644
--- a/pkgs/development/misc/resholve/default.nix
+++ b/pkgs/development/misc/resholve/default.nix
@@ -1,16 +1,28 @@
-{ callPackage
-, deNeuteredPy27OnlyForResholve
+{ lib
+, pkgs
+, pkgsBuildHost
 , ...
 }:
 
 let
+  python27' = (pkgsBuildHost.python27.overrideAttrs (old:
+    {
+      # Overriding `meta.knownVulnerabilities` here, see #201859 for why it exists
+      # In resholve case this will not be a security issue,
+      # since the code should always run inside nix sandbox (when enabled)
+      meta = old.meta // { knownVulnerabilities = [ ]; };
+    }
+  )).override {
+    self = python27';
+    pkgsBuildHost = pkgsBuildHost // { python27 = python27'; };
+  };
+  callPackage = lib.callPackageWith (pkgs // { python27 = python27'; });
   source = callPackage ./source.nix { };
-  deps = callPackage ./deps.nix { python27 = deNeuteredPy27OnlyForResholve; };
+  deps = callPackage ./deps.nix { };
 in
 rec {
   # resholve itself
   resholve = callPackage ./resholve.nix {
-    python27 = deNeuteredPy27OnlyForResholve;
     inherit (source) rSrc version;
     inherit (deps.oil) oildev;
     inherit resholve-utils;
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 86cde56da9c..e24d1e9130d 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -15868,10 +15868,6 @@ with pkgs;
     pythonAttr = "python2Full";
     x11Support = true;
   };
-  deNeuteredPy27OnlyForResholve = (python27.overrideAttrs(old: { meta = old.meta // { knownVulnerabilities = []; };})).override {
-    self = deNeuteredPy27OnlyForResholve;
-    pythonAttr = "deNeuteredPy27OnlyForResholve";
-  };
   python27Full = python27.override {
     self = python27Full;
     pythonAttr = "python27Full";
```

**Comment by abathur** [comment]
@cole-h yes, sorry; meant to but was in a hurry to get it posted before leaving for dinner.

**Comment by abathur** [comment]
force-pushed; removed unwanted commit and square w/ 2nd patch by @thiagokokada

**Review by thiagokokada — APPROVED** [review]
Think this is ready to merge.

Any reason to have to keep this as draft?

**Comment by abathur** [comment]
@thiagokokada marked ready for review

**Comment by thiagokokada** [comment]
An alternative way to do this:

```patch
diff --git a/pkgs/development/misc/resholve/default.nix b/pkgs/development/misc/resholve/default.nix
index 55c0c7ad046..6b8e7199418 100644
--- a/pkgs/development/misc/resholve/default.nix
+++ b/pkgs/development/misc/resholve/default.nix
@@ -1,22 +1,19 @@
 { lib
-, pkgs
 , pkgsBuildHost
+, system
 , ...
 }:
 
 let
-  python27' = (pkgsBuildHost.python27.overrideAttrs (old:
-    {
-      # Overriding `meta.knownVulnerabilities` here, see #201859 for why it exists
-      # In resholve case this should not be a security issue,
-      # since the code will run inside nix sandbox (when enabled)
-      meta = old.meta // { knownVulnerabilities = [ ]; };
-    }
-  )).override {
-    self = python27';
-    pkgsBuildHost = pkgsBuildHost // { python27 = python27'; };
+  pkgs = import ../../../.. {
+    inherit system;
+    # Allow python27 with known security issues only for resholve,
+    # see issue #201859 for the reasoning
+    # In resholve case this should not be a security issue,
+    # since it will only be used during build, not runtime
+    config.permittedInsecurePackages = [ pkgsBuildHost.python27.name ];
   };
-  callPackage = lib.callPackageWith (pkgs // { python27 = python27'; });
+  callPackage = lib.callPackageWith pkgs;
   source = callPackage ./source.nix { };
   deps = callPackage ./deps.nix { };
 in
```

Not sure which one is better. This one does not override the package and explicitly add `python27` to the `permittedInsecurePackages` allow list. We are re-evaluating `nixpkgs` though, but this shouldn't be a huge issue (some packages inside the bootstrap also does the same).

**Comment by abathur** [comment]
Force-push to adopt Thiago's more recent suggestion.

Realized after pushing that this approach probably makes https://github.com/NixOS/nixpkgs/pull/204023/commits/afce69f2ba553bbe548c4224ac037144dc4f63e9 not-technically-necessary, though it does better homogenize the python27 package with python3x. Want me to drop that commit?

**Comment by abathur** [comment]
Force-pushed to drop python27 overridability change in https://github.com/NixOS/nixpkgs/commit/afce69f2ba553bbe548c4224ac037144dc4f63e9.

**Review by thiagokokada — APPROVED** [review]

**Comment by dotlambda** [comment]
What's the timeline for getting rid of this band-aid?

**Comment by ajs124** [comment]
`system` is an alias, so this breaks eval without aliases enabled

**Comment by thiagokokada** [comment]
> `system` is an alias, so this breaks eval without aliases enabled

We can use `stdenv.hostPlatform.system` instead. Would this work?

**Comment by dotlambda** [comment]
Can someone explain the rationale behind this? In which sense would resholve not be affected by Python 2 vulnerabilities? It's also intended to be used outside of Nixpkgs and the Nix sandbox.

**Comment by thiagokokada** [comment]
> Can someone explain the rationale behind this? In which sense would resholve not be affected by Python 2 vulnerabilities? It's also intended to be used outside of Nixpkgs and the Nix sandbox.

The sandbox comment was a mistake, the point is not the sandbox at all. Resholve itself will never connect to internet for anything.

Also, this is only a build time dependency. It will not be included at runtime. If build time out-of-date dependencies are an issue, what about our derivations that depend on older versions of GCC? Heck, we probably depends on some older unsupported GCC for bootstrapping. Is this also an issue?

**Comment by thiagokokada** [comment]
BTW, we use unsupported versions of compilers for a lots of things here in nixpkgs. I just recently discover we do the same for OpenJDK (we will use the `openjdk13` to bootstrap `openjdk14`, same for `openjdk15`, and this continues until we get to a modern supported version). This was never an issue, and `openjdk12` has way more known security issues then Python 2.7 ever had.

Not sure why Python 2 gets a different treatment. We are basically doing the same thing as we do with `openjdk` here, marking outdated versions as EOL and only allowing it for build purposes.

**Comment by dotlambda** [comment]
> Not sure why Python 2 gets a different treatment. We are basically doing the same thing as we do with `openjdk` here, marking outdated versions as EOL and only allowing it for build purposes.

Resholve uses Python 2 at runtime.
And its [readme](https://github.com/abathur/resholve#diy) explains how to use resholve outside of Nixpkgs. In that case, the Python 2 code could even be run on untrusted input.

**Comment by dotlambda** [comment]
I guess it would be okay to remove resholve as a top-level attribute and keep it for "bootstrapping" some command line scripts only. That's what OpenJDK does.

**Comment by bobrik** [comment]
Looks like it broke msmtp (I'm on `aarch64-darwin`):

```
$ nix-env --install --attr nixpkgs.msmtp --dry-run
(dry run; not doing anything)
installing 'msmtp-1.8.22'
error: pyyaml-6.0 not supported for interpreter python2.7
(use '--show-trace' to show detailed location information)
```

**Comment by thiagokokada** [comment]
> Looks like it broke msmtp (I'm on `aarch64-darwin`):
> 
> ```
> $ nix-env --install --attr nixpkgs.msmtp --dry-run
> (dry run; not doing anything)
> installing 'msmtp-1.8.22'
> error: pyyaml-6.0 not supported for interpreter python2.7
> (use '--show-trace' to show detailed location information)
> ```

Not related to this PR.

**Comment by bobrik** [comment]
Actual change that broke msmtp: https://github.com/NixOS/nixpkgs/pull/204405#issuecomment-1344655876.

**Comment by ahmedelgabri** [comment]
Looks like it broke more packages, see also #205438

**Comment by dotlambda** [comment]
So nobody has an argument for this special treatment?

**Comment by thiagokokada** [comment]
> So nobody has an argument for this special treatment?

What do you mean? I did argue above.

**Comment by dotlambda** [comment]
> > So nobody has an argument for this special treatment?
> 
> What do you mean? I did argue above.

You argued for resholve as a build time dependency, but there's currently no indication that's the only way resholve is supposed to be used.

**Comment by jonringer** [comment]
I feel like this is an anti-pattern, essentially this overrides what the user provided `config` to enable a certain scenario. 

Python2 should really be getting phased out, not adding complexity to nixpkgs. I realize this is a personally maintained project from a contributor, but python2 well be soon 3 years EOL....

I guess I would like to see the python3 migration (or migration to another less-maintainer-and-downstream-burdensome language) be prioritized.

**Comment by thiagokokada** [comment]
> > > So nobody has an argument for this special treatment?
> > 
> > What do you mean? I did argue above.
> 
> You argued for resholve as a build time dependency, but there's currently no indication that's the only way resholve is supposed to be used.

IMO `resholve.mkDerivation` is the only real use of resholve, while the derivation itself is mostly used for debugging purposes. However @abathur may have more information about it.

If you feel really strong about it, we probably can remove the exposure of the derivation in `all-packages.nix`. This should take care of the issue.

**Comment by thiagokokada** [comment]
> I feel like this is an anti-pattern, essentially this overrides what the user provided `config` to enable a certain scenario. 
> 
> Python2 should really be getting phased out, not adding complexity to nixpkgs. I realize this is a personally maintained project from a contributor, but python2 well be soon 3 years EOL....
> 
> I guess I would like to see the python3 migration (or migration to another less-maintainer-and-downstream-burdensome language) be prioritized. 

I think it is fair to have this as a migration path, and give time to the maintainer to move it to Python 3. It is not a trivial task considering that `oilshell` (the project that `resholve` uses as a parser) is written in Python 2 with no plans to migrate to Python 3.

**Comment by dotlambda** [comment]
> I think it is fair to have this as a migration path, and give time to the maintainer to move it to Python 3.

It's been three years.

**Comment by thiagokokada** [comment]
> > I think it is fair to have this as a migration path, and give time to the maintainer to move it to Python 3.
> 
> It's been almost three years.

I talked with @abathur in private. `oildev` used to be much more unstable before, so trying to create a Python 3 API wrapping `oildev` own's Python 2 implementation would be very fragile.

Nowadays this looks much more promising thanks to the stabilization of the code and the amount of tests that increased in the recent years.

Still this is a lot of work, so we should give time to the maintainer to do it's job.

File: pkgs/development/misc/resholve/default.nix [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
 , ...
 }:
 
 let
+  pkgs = import ../../../.. {
+    inherit system;
+    # Allow python27 with known security issues only for resholve,
+    # see issue #201859 for the reasoning
+    # In resholve case this should not be a security issue,
+    # since it will only be used during build, not runtime
```
[the remark this answers is not in the document]
Comment (thiagokokada) [reply pkgs/development/misc/resholve/default.nix:13]: I am going to work in something later. Probably going to throw in `resholve` and expose only `resholve-utils`.

This way is pretty much impossible for someone to use resholve at runtime.

[the remark this answers is not in the document]
Comment (thiagokokada) [reply pkgs/development/misc/resholve/default.nix:13]: Done: https://github.com/NixOS/nixpkgs/pull/210078

**Pull request merged**